De-randomize assertion ordering in plain report

xbauch opened the pull request:
specifically order also based on function-name and only compare line number if they differ.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

chrisr-diffblue commented on src/goto-checker/report_util.cpp:
```diff
@@ -198,7 +198,11 @@ get_sorted_properties(const propertiest &properties)
       const auto &p2 = pit2->second.pc->source_location;
       if(p1.get_file() != p2.get_file())
         return id2string(p1.get_file()) < id2string(p2.get_file());
-      else if(!p1.get_line().empty() && !p2.get_line().empty())
```
Github won't let me comment on the correct lines, but a few lines above this there is a comment that describes the ordering - I think that needs updated to reflect your changes.

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-checker/report_util.cpp:
```diff
@@ -198,7 +199,11 @@ get_sorted_properties(const propertiest &properties)
+        p1.get_line() != p2.get_line())
         return std::stoul(id2string(p1.get_line())) <
                std::stoul(id2string(p2.get_line()));
       else
```
I think it’d probably be better to sort ids numerically (by the number at the end) rather than lexicographically here

xbauch replied:
Done.

hannes-steffenhagen-diffblue reviewed (commented):
LGTM, but I imagine some tests will have to be updated?

chrisr-diffblue reviewed (approved):
Looks OK to me, pending un-drafting and passing CI, of course :-)

martin-cs reviewed (approved):
Seems reasonable and removing this unnecessary platform non-determinism seems good.  I am a little concerned about what this might break in test suites but I guess we will see.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 26f62822dcd7c51379b2748dc05dbf07d36aaba4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/140982791

petr-bauch committed:
De-randomize assertion ordering in plain report

specifically order also based on function-name and only compare line number if
they differ. Moreover, we extend the id comparison to be alphabetical for the
assertion name and numerical for the number part.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ed498df3ab2a86fd042bdd89d5777a7f9bad95f4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/141450413
